// miter/dist/metadata/server/orm.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const injectable_decorator_1 = require("../../decorators/services/injectable.decorator");
const server_1 = require("./server");
let OrmMetadata = OrmMetadata_1 = class OrmMetadata {
    constructor(_meta) {
        this._meta = _meta;
    }
    get originalMeta() {
        return this._meta;
    }
    get enabled() {
        if (!this._meta || typeof this._meta.enabled === 'undefined')
            return true;
        return !!this._meta.enabled;
    }
    get models() {
        return (this._meta && this._meta.models) || [];
    }
    get recreate() {
        return !!(this._meta && this._meta.recreate);
    }
};
OrmMetadata = OrmMetadata_1 = __decorate([
    injectable_decorator_1.Injectable({
        provide: {
            useCallback: function (meta) {
                let ormMeta = meta.originalMeta.orm;
                return new OrmMetadata_1(ormMeta);
            },
            deps: [server_1.ServerMetadata],
            cache: true
        }
    }),
    __metadata("design:paramtypes", [Object])
], OrmMetadata);
exports.OrmMetadata = OrmMetadata;
var OrmMetadata_1;
