// miter/dist/metadata/server/router.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const injectable_decorator_1 = require("../../decorators/services/injectable.decorator");
const server_1 = require("./server");
let RouterMetadata = RouterMetadata_1 = class RouterMetadata {
    constructor(_meta) {
        this._meta = _meta;
    }
    get originalMeta() {
        return this._meta;
    }
    get path() {
        return this._meta.path || '';
    }
    get middleware() {
        return this._meta.middleware || [];
    }
    get policies() {
        return this._meta.policies || [];
    }
    get controllers() {
        return this._meta.controllers || [];
    }
};
RouterMetadata = RouterMetadata_1 = __decorate([
    injectable_decorator_1.Injectable({
        provide: {
            useCallback: function (meta) {
                let routerMeta = meta.originalMeta.router;
                return !!routerMeta ? new RouterMetadata_1(routerMeta) : null;
            },
            deps: [server_1.ServerMetadata],
            cache: true
        }
    }),
    __metadata("design:paramtypes", [Object])
], RouterMetadata);
exports.RouterMetadata = RouterMetadata;
var RouterMetadata_1;
