// miter/dist/metadata/server/server.js
"use strict";
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
Object.defineProperty(exports, "__esModule", { value: true });
const injectable_decorator_1 = require("../../decorators/services/injectable.decorator");
let ServerMetadata = ServerMetadata_1 = class ServerMetadata {
    constructor(_meta) {
        this._meta = _meta;
    }
    get originalMeta() {
        return this._meta;
    }
    get name() {
        return this._meta.name;
    }
    get port() {
        return this._meta.port || 80;
    }
    get allowCrossOrigin() {
        return this._meta.allowCrossOrigin || false;
    }
    get services() {
        return this._meta.services || [];
    }
    get inject() {
        return this._meta.inject || [];
    }
    get logLevel() {
        return this._meta.logLevel || 'warn';
    }
    get debugBreakpoint() {
        return this._meta.debugBreakpoint || (() => void (0));
    }
};
ServerMetadata = ServerMetadata_1 = __decorate([
    injectable_decorator_1.Injectable({
        provide: {
            useCallback: function (meta) {
                return new ServerMetadata_1(meta);
            },
            deps: ['server-meta'],
            cache: true
        }
    }),
    __metadata("design:paramtypes", [Object])
], ServerMetadata);
exports.ServerMetadata = ServerMetadata;
var ServerMetadata_1;
