// miter/dist/metadata/orm/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var Types;
(function (Types) {
    Types[Types["string"] = 0] = "string";
    Types[Types["text"] = 1] = "text";
    Types[Types["integer"] = 2] = "integer";
    Types[Types["bigint"] = 3] = "bigint";
    Types[Types["float"] = 4] = "float";
    Types[Types["real"] = 5] = "real";
    Types[Types["double"] = 6] = "double";
    Types[Types["decimal"] = 7] = "decimal";
    Types[Types["date"] = 8] = "date";
    Types[Types["dateonly"] = 9] = "dateonly";
    Types[Types["boolean"] = 10] = "boolean";
    Types[Types["enum"] = 11] = "enum";
    Types[Types["uuid"] = 12] = "uuid";
    Types[Types["virtual"] = 13] = "virtual";
    Types[Types["array"] = 14] = "array";
    Types[Types["json"] = 15] = "json";
    Types[Types["jsonb"] = 16] = "jsonb";
    Types[Types["blob"] = 17] = "blob";
    Types[Types["range"] = 18] = "range";
    Types[Types["geometry"] = 19] = "geometry";
})(Types = exports.Types || (exports.Types = {}));
